Add documentation for Azure OpenAI

shivam-51 opened the pull request:
Thank you for helping make the Dapr documentation better!

**Please follow this checklist before submitting:**
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://docs.dapr.io/contributing/contributing-overview/#developer-certificate-of-origin-signing-your-work))
- [x] [Read the contribution guide](https://docs.dapr.io/contributing/docs-contrib/contributing-docs/)
- [ ] Commands include options for Linux, MacOS, and Windows within codetabs
- [x] New file and folder names are globally unique
- [x] Page references use shortcodes instead of markdown or URL links
- [ ] Images use HTML style and have alternative text
- [x] Places where multiple code/command options are given have codetabs

In addition, please fill out the following to help reviewers understand this pull request:

## Description

Add documentation for Azure OpenAI output binding
Refer to this PR for more details https://github.com/dapr/components-contrib/pull/2965

## Issue reference

https://github.com/dapr/components-contrib/issues/2716

shivam-51 committed:
Add documentation for Azure OpenAI

Signed-off-by: Shivam Kumar Singh <shivamhere247@gmail.com>

shivam-51 commented:
@ItalyPaleAle Please have a look.

ItalyPaleAle commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,236 @@
+
+## Component format
+
+To setup OpenAI binding create a component of type `bindings.azure.openai`. See [this guide]({{< ref "howto-bindings.md#1-create-a-binding" >}}) on how to create and apply a binding configuration.
```
It's important we always call this "Azure OpenAI"

```suggestion
To setup an Azure OpenAI binding create a component of type `bindings.azure.openai`. See [this guide]({{< ref "howto-bindings.md#1-create-a-binding" >}}) on how to create and apply a binding configuration.
```

ItalyPaleAle commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,236 @@
+## Component format
+
+To setup OpenAI binding create a component of type `bindings.azure.openai`. See [this guide]({{< ref "howto-bindings.md#1-create-a-binding" >}}) on how to create and apply a binding configuration.
+See [this](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/overview/) for the documentation for Azure OpenAI Service.
```
```suggestion
See [this](https://learn.microsoft.com/azure/cognitive-services/openai/overview/) for the documentation for Azure OpenAI Service.
```

ItalyPaleAle commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,236 @@
+
+| Field              | Required | Binding support | Details | Example |
+|--------------------|:--------:|--------|---------|---------|
+| `endpoint` | Y | Input/Output | Azure OpenAI service endpoint URL. | `"https://myopenai.openai.azure.com"` |
+| `apiKey` | Y* | Input/Output | The access key of the Azure OpenAI service. Only required when not using Azure AD authentication. | `"1234567890abcdef"` |
```
```suggestion
| `endpoint` | Y | Output | Azure OpenAI service endpoint URL. | `"https://myopenai.openai.azure.com"` |
| `apiKey` | Y* | Output | The access key of the Azure OpenAI service. Only required when not using Azure AD authentication. | `"1234567890abcdef"` |
```

ItalyPaleAle commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,236 @@
+- [How-To: Trigger application with input binding]({{< ref howto-triggers.md >}})
+- [How-To: Use bindings to interface with external resources]({{< ref howto-bindings.md >}})
+- [Bindings API reference]({{< ref bindings_api.md >}})
+- [Azure OpenAI Rest examples](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference)
```
```suggestion
- [Azure OpenAI Rest examples](https://learn.microsoft.com/azure/cognitive-services/openai/reference)
```

ItalyPaleAle reviewed (commented):
Thanks for creating these docs!

shivam-51 committed:
Add documentation for Azure OpenAI output binding

Signed-off-by: Shivam Kumar Singh <shivamhere247@gmail.com>

Resolve comments

Update daprdocs/content/en/reference/components-reference/supported-bindings/openai.md

Co-authored-by: Mark Fussell <markfussell@gmail.com>
Signed-off-by: Shivam Kumar Singh <shivamhere247@gmail.com>

Update daprdocs/content/en/reference/components-reference/supported-bindings/openai.md

Co-authored-by: Mark Fussell <markfussell@gmail.com>
Signed-off-by: Shivam Kumar Singh <shivamhere247@gmail.com>

Resolve comments

Signed-off-by: Shivam Kumar Singh <shivamhere247@gmail.com>

msfussell commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,236 @@
+  type: bindings.azure.openai
+  version: v1
+  metadata:
+  - name: apiKey
```
```suggestion
  - name: apiKey # Required
```

shivam-51 replied:
Done.

msfussell commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
````diff
@@ -0,0 +1,236 @@
+    value: "my-model"
+```
+{{% alert title="Warning" color="warning" %}}
+The above example uses apiKey as plain strings. It is recommended to use a secret store for the secrets as described [here]({{< ref component-secrets.md >}}).
````
```suggestion
The above example uses `apiKey` as  a plain string. It is recommended to use a secret store for the secrets as described [here]({{< ref component-secrets.md >}}).
```

msfussell commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,236 @@
+
+- `prompt` - string that specifies the prompt to generate completions for.
+- `maxTokens` - (optional) defines the max number of tokens to generate. Defaults to 16 for completion API.
+- `temperature` - (optional) defines the sampling temperature between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic. Defaults to 1.0 for completion API.
```
```suggestion
- `temperature` - (optional) defines the sampling temperature between 0 and 2. Higher values like 0.8 make the output more random, while lower values like 0.2 make it more focused and deterministic. Defaults to 1.0 for completion API.
```

msfussell commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,236 @@
+
+#### Response
+
+The response body will contain the following JSON:
```
```suggestion
The response body contains the following JSON:
```

msfussell commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,236 @@
+  - `role` - string that specifies the role of the message. Can be either `user`, `system` or `assistant`.
+  - `message` - string that specifies the conversation message for the role.
+- `maxTokens` - (optional) defines the max number of tokens to generate. Defaults to 16 for the chat completion API.
+- `temperature` - (optional) defines the sampling temperature between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic. Defaults to 1.0 for the chat completion API.
```
```suggestion
- `temperature` - (optional) defines the sampling temperature between 0 and 2. Higher values like 0.8 make the output more random, while lower values like 0.2 make it more focused and deterministic. Defaults to 1.0 for the chat completion API.
```

msfussell commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,236 @@
+
+#### Response
+
+The response body will contain the following JSON:
```
```suggestion
The response body contains the following JSON:
```

msfussell commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,236 @@
+| `endpoint` | Y | Output | Azure OpenAI service endpoint URL. | `"https://myopenai.openai.azure.com"` |
+| `apiKey` | Y* | Output | The access key of the Azure OpenAI service. Only required when not using Azure AD authentication. | `"1234567890abcdef"` |
+| `deploymentId` | Y | Output | The name of the model deployment. | `"my-model"` |
+
```
Can you also include the Azure AAD authentication fields into this table, as shown in the example below.  You can take the same descriptions from here https://docs.dapr.io/reference/components-reference/supported-bindings/eventgrid/

shivam-51 replied:
Done. Please have a look.

msfussell reviewed (commented):
Great documentation! Left a few comments

shivam-51 committed:
Merge branch 'v1.11' into add-documentation-for-azure-openai

shivam-51 commented:
@msfussell Please have a look.

shivam-51 committed:
Update links

shivam-51 commented:
@mukundansundar Please have a look.

msfussell reviewed (approved):
LGTM

hhunter-ms commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,239 @@
+- `presencePenalty` - (optional) Number between -2.0 and 2.0. Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics. Defaults to 0.0 for completion API.
+- `frequencyPenalty` - (optional) Number between -2.0 and 2.0. Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim. Defaults to 0.0 for completion API.
+
+Read more about the importance and usage of these parameters in the [Azure OpenAI API documentation](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference).
```
```suggestion
Read more about the importance and usage of these parameters in the [Azure OpenAI API documentation](https://learn.microsoft.com/azure/ai-services/openai/reference).
```

hhunter-ms commented on daprdocs/content/en/reference/components-reference/supported-bindings/openai.md:
```diff
@@ -0,0 +1,239 @@
+- [How-To: Trigger application with input binding]({{< ref howto-triggers.md >}})
+- [How-To: Use bindings to interface with external resources]({{< ref howto-bindings.md >}})
+- [Bindings API reference]({{< ref bindings_api.md >}})
+- [Azure OpenAI Rest examples](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference)
```
```suggestion
- [Azure OpenAI Rest examples](https://learn.microsoft.com/azure/ai-services/openai/reference)
```

hhunter-ms committed:
Update daprdocs/content/en/reference/components-reference/supported-bindings/openai.md

Signed-off-by: Hannah Hunter <94493363+hhunter-ms@users.noreply.github.com>

hhunter-ms committed:
Update daprdocs/content/en/reference/components-reference/supported-bindings/openai.md

Signed-off-by: Hannah Hunter <94493363+hhunter-ms@users.noreply.github.com>
